### PR TITLE
Added `n_nonzero_elements` to PETSc `BlockSparseMatrix`.

### PR DESCRIPTION
**Pull request by marcfehling**
I noticed that this particular function was missing.

Copied the doc from the [PETSc MatrixBase](https://www.dealii.org/developer/doxygen/deal.II/classPETScWrappers_1_1MatrixBase.html#adfcc9652ade95e371d4191e28461eeb8) class, and the implementation from the [Trilinos BlockSparseMatrix](https://github.com/dealii/dealii/blob/6c065836a1ff9259393f6e851c60b20ec0cddb13/source/lac/trilinos_block_sparse_matrix.cc#L235-L244) class.

### CODE REVIEW

**Comment by marcfehling** [comment]
I have no idea what doxygen complains about.
```
/home/runner/work/dealii/dealii/source/lac/petsc_parallel_block_sparse_matrix.cc:36: warning: no uniquely matching class member found for 
  void BlockSparseMatrix< number >::reinit(unsigned int n_block_rows, const size_type n_block_columns)
```
The function in line 36 of that file is *not* a template function, yet doxygen claims a template `number` exists.

Maybe it gets confused with one of the other `BlockSparseMatrix` classes, in particular [this one](https://www.dealii.org/developer/doxygen/deal.II/classBlockSparseMatrix.html).

File: source/lac/petsc_parallel_block_sparse_matrix.cc [hunk 1]
```diff
@@ -33,8 +33,9 @@ namespace PETScWrappers
 
 
     void
-    BlockSparseMatrix::reinit(const size_type n_block_rows,
-                              const size_type n_block_columns)
+    BlockSparseMatrix::reinit(
+      const BlockSparseMatrix::size_type n_block_rows,
+      const BlockSparseMatrix::size_type n_block_columns)
```
Comment (bangerth) [inline source/lac/petsc_parallel_block_sparse_matrix.cc:38]: Just undo this change. Once you're past the opening `(`, you're in scope for the class, and you don't need to qualify `size_type` with the class name any more.

**Comment by bangerth** [comment]
Is there an existing test one could easily adapt for this function? Maybe from the Trilinos classes?

File: source/lac/petsc_parallel_block_sparse_matrix.cc (hunk 1 above)
Comment (marcfehling) [reply source/lac/petsc_parallel_block_sparse_matrix.cc:38]: I will remove this change, as doxygen still complains. But it looks like it gives us more insight what's going wrong.
```
/home/runner/work/dealii/dealii/source/lac/petsc_parallel_block_sparse_matrix.cc:36: warning: no uniquely matching class member found for 
  void BlockSparseMatrix< number >::reinit(unsigned int n_block_rows, const BlockSparseMatrix::size_type n_block_columns)
```
`n_block_rows` is an `unsigned int`, but `n_block_columns` is a `size_type`?

**Comment by marcfehling** [comment]
> Is there an existing test one could easily adapt for this function? Maybe from the Trilinos classes?

It looks like we don't check for `TrilinosWrappers::BlockSparseMatrix::n_nonzero_elements` in our testsuite. I will create tests for either class based upon [tests/lac/block_matrices](https://github.com/dealii/dealii/blob/master/tests/lac/block_matrices.cc).

File: source/lac/petsc_parallel_block_sparse_matrix.cc (hunk 1 above)
Comment (marcfehling) [reply source/lac/petsc_parallel_block_sparse_matrix.cc]: I have a strong feeling that `n_block_rows` gets confused with the member function [`n_block_rows`](https://www.dealii.org/developer/doxygen/deal.II/classBlockMatrixBase.html#a2478164707bf2932ad32ddd27babfc00).

There is also a function `n_block_cols`, but the argument in the `reinit` function is `n_block_columns` so it is not ambiguous.

Comment (marcfehling) [reply source/lac/petsc_parallel_block_sparse_matrix.cc]: The `Trilinos` equivalent has simply been hidden from `doxygen` in #9880. I will do the same here.
https://github.com/dealii/dealii/blob/6c065836a1ff9259393f6e851c60b20ec0cddb13/source/lac/trilinos_block_sparse_matrix.cc#L41-L67

**Comment by marcfehling** [comment]
> Is there an existing test one could easily adapt for this function? Maybe from the Trilinos classes?

Created simple tests for both PETSc and Trilinos sparse matrices. Started with the `lac/block_matrices` test which didn't work with Trilinos because of #13491. So I created smaller, much simpler tests from scratch.

Ready for review.

**Review by bangerth — APPROVED** [review]
